### ISSUE TITLE
Scale/create node pools to size 0 for auxiliary node pools via AKS API

### ISSUE THREAD

**Issue by BoSoeborgPetersen** [Azure/AKS#1050]
It should be possible to scale a nodepool to zero through the Azure Cli, not just through Azure Portal.

**What happened**:
Received error ("Can't scale down to 0 node.")

**What you expected to happen**:
Scale down to zero nodes.

**How to reproduce it (as minimally and precisely as possible)**:
"az aks nodepool scale -g <group_name> --cluster-name <cluster-name> --node-count 0 -n winvms”

**Anything else we need to know?**:
Azure CLI version 2.0.66

**Environment**:
Kubernetes version (use `kubectl version`): 1.14.0

**Comment by ams0** [comment]
It's possible operating on the VMSS scale directly:

```
az vmss scale --name MyScaleSet --new-capacity 0 --resource-group MyResourceGroup
```

**Comment by BoSoeborgPetersen** [comment]
Good to know, but it should still be fixed.

**Comment by jluk** [comment]
Node pools have a minimum size of 1 within AKS today during preview. We plan to support size zero node pools but this requires additional work - FYI @xizhamsft

**Comment by BoSoeborgPetersen** [comment]
Then why is it possible to scale to 0 nodes using the portal (on the AKS resource)?

**Comment by jluk** [comment]
@BoSoeborgPetersen can you explain the steps you're using to do that via portal? You can scale to 0 from the VMSS clients, but that isn't supported given it's IaaS held outside of AKS as @ams0 mentioned. 

From AKS portal you have a min required size of 1 node for node pools. If you only have one node pool you need at least one node.

![image](https://user-images.githubusercontent.com/3340350/59866713-1cfcd600-9341-11e9-9436-b967e1c798b1.png)

**Comment by local-master** [comment]
@jluk This is scaling through "Kubernetes service" in Azure Portal. We are talking about Scaling one nodepool from the VMSS ( Virtual machine scale sets ) It is possible to scale pool to 0 nodes over the Azure Portal.

**Comment by jluk** [comment]
@local-master understood, that screenshot is on the AKS blade which calls the AKS API to scale. To clarify, the only way to scale a VMSS to 0 is via the VMSS blade which calls the VMSS API directly and it is not supported.

As you'll see in the docs it only uses the AKS clients, not the VMSS clients which use a separate API.

Renaming the issue to be more accurate which is something we want to support via AKS.

**Comment by victornoel** [comment]
For the record I tried the following (tl;dr it broke my cluster):
- I added a second node pool to my cluster
- I cordoned the sole node of the primary node pool then drained it
- everything went as expected on the other nodes
- I scaled to zero via the VMSS portal and the cluster simply stopped answering on the public ip that was used by the ingress service configured via `LoadBalancerIP` even though kubernetes said everything was ok
- After scaling back to 1 node via the VMSS portal, the cluster started answering again on the public ip

**Comment by jluk** [comment]
@victornoel late reply, but taking the VMSS path sends the cluster out of support since it's an out of band operation outside of AKS. As you witnessed (or as I read your repro) you will always need nodes in your default cluster as that is where critical system pods end up.

We're still tracking to support this scale to 0 for auxiliary node pools, targeting before end of year.

**Comment by yaariyuval** [comment]
@jluk any update on this issue now that multiple node pools are GA? Thanks in advance!

**Comment by jluk** [comment]
@yaariyuval this work is still in planning, we expect to complete it but was not a GA blocker for the all-up feature. Be on the lookout in the coming months for us to close on this one.

**Comment by wildermuthn** [comment]
This would be really helpful for GPU node pools for Kubeflow pipelines that aren't being run 24-7.

**Comment by nirp123** [comment]
+1
AKS multi node pools without node count = 0 is useless.
please hurry up with this feature

**Comment by mboutet** [comment]
+1 Absolutely need this feature, especially for GPU node.

**Comment by Robert-Stam** [comment]
+1 This would help me us well

**Comment by michael-burt** [comment]
This is a very important feature for many different use cases. Any update on implementation? AKS is the only manager Kubernetes offering that does not support this among the top cloud providers.

**Comment by Shaked** [comment]
Any idea when this will be available? 

Also, what would be the best solution until this feature is released? We would like to make sure we are not wasting money on unused resources while our GPUs are not being used. 

Thank you
Shaked

**Comment by focom** [comment]
What is the ETA on this feature? I want to make wise choice on what hack i should use to circumvent this.

**Comment by jay-bhambhani** [comment]
hi there! would also love to implement this feature to save us some cash :smile:

**Comment by jluk** [comment]
Thanks everyone for the excitement and feedback on this feature! 

We've begun the work to remove blockers for secondary node pools of size zero. This starts with ensuring the requirements of a cluster can be validated on a subset of pools in a cluster, instead of all the pools individually. As some may be aware, options like Azure CNI require a minimum set of nodes per pool. We have this issue in-flight, which once completed will unblock support of a size 0 aux pool. (https://github.com/Azure/AKS/issues/1309)

We plan to enable this within CY20Q2, as we make progress we will include more details as we go.

**Comment by tmbgreaves** [comment]
Adding my voice of interest to this, thanks @jluk for your team's continued work to improve AKS! 

Our use cases: Jupyterhub instances on AKS for teaching ML courses where we want to minimise spend on GPU nodes for the majority of the time when students aren't working on the hub, but want to keep the hub online with a few CPU nodes so students can access files pre/post class. Also hands-on conference sessions where we want a Jupyterhub instance sitting on tick-over and just needing a GPU pool scale-up before a session starts.

**Comment by mlushpenko** [comment]
Hehe, @tmbgreaves we had exactly the same use-case in terms of tech, just we had customer users instead of students

**Comment by jluk** [comment]
Thanks for sharing, these are 100% aligned with our goals on this feature. We are closing down on the final stages to support this and should have good updates in the next 2 weeks.

**Comment by damienpontifex** [comment]
Release notes seem that this is either available or becoming available with the latest release rollout https://github.com/Azure/AKS/releases/tag/2020-04-13

**Comment by jluk** [comment]
Correct, this should be available everywhere now for clusters with at least one system pool. Learn more at aka.ms/aks/nodepool/mode thanks for the feedback!

**Comment by Robert-Stam** [comment]
Is the maximum number of nodepool (which was first 7 and the last few months 10) dropped as well, it seems to be removed from the restrictions, however not explicitly mentioned in the release notes?

**Comment by Robert-Stam** [comment]
> Correct, this should be available everywhere now for clusters with at least one system pool. Learn more at aka.ms/aks/nodepool/mode thanks for the feedback!

@jluk Is the maximum number of nodepools (which was first 7 and the last few months 10) dropped as well, it seems to be removed from the restrictions, however not explicitly mentioned in the release notes?

**Comment by jluk** [comment]
The maximum count of nodepools today is still 10 in a single cluster, thanks for raising will revisit that doc.

**Comment by Robert-Stam** [comment]
> The maximum count of nodepools today is still 10 in a single cluster, thanks for raising will revisit that doc.

@jluk Will this be increased in the near future, we have hard times with this low number?

We are running AKS clusters with:
 - multiple availability zones (3)
 - multiple VM-scalesets (aks nodepools) currently up-to 7
 - with the cluster auto-scaler enabled (1-30 nodes/np)
 - using persistent storage / PVC (from multiple AZ, dependent on where the pod is launched the first time) 
 - installed Kured for rebooting nodes which need to (for updates)
 
We run into the issue that Kured is cordoning/draining nodes and the auto-scaler is creating new ones, however not always in the same AZ.
This results in pods that cannot mount it's volume anymore.
As workaround I found this: https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/FAQ.md#i-have-a-couple-of-pending-pods-but-there-was-no-scale-up
This implies that we add more node-pools, however we cannot add more than 10 ATM.

It would be great if we could (and AKS can check at runtime if we respect the limit of 1000 nodes in total).

Thanks in advance!

**Comment by jluk** [comment]
Thanks @robert-stam for the feedback, let's use #1570 to have this specific discussion. We are always open to evaluating limits like that, but I'm interested in the root issue caused by kured/pvs.

**Comment by Robert-Stam** [comment]
> Thanks @Robert-Stam for the feedback, let's use #1570 to have this specific discussion. We are always open to evaluating limits like that, but I'm interested in the root issue caused by kured/pvs.

Thanks, added my question there.

**Comment by 0GiS0** [comment]
Hi all,

I wrote an Azure Function to try to scale to 0 using the following code and **It works perfectly**:

`var msRestNodeAuth = require('@azure/ms-rest-nodeauth');
const msRestAzure = require('@azure/ms-rest-azure-js');

module.exports = async function (context, req) {

    var creds = await msRestNodeAuth.loginWithServicePrincipalSecret(
        process.env.CLIENT_ID,
        process.env.SECRET,
        process.env.TENANT_ID
    );

    var client = new msRestAzure.AzureServiceClient(creds);

    var options = {
        method: 'PUT',
        url: `https://management.azure.com/subscriptions/${req.body.subscriptionId}/resourceGroups/${req.body.resourceGroupName}/providers/Microsoft.ContainerService/managedClusters/${req.body.resourceName}/agentPools/${req.body.agentPoolName}?api-version=2020-03-01`,
        body: {
            "properties.count": req.body.count
        }
    };

    try {
        var response = await client.sendRequest(options);
        context.log(response);
        context.log(`status code: ${response.status}`);

        context.res = {
            body: `Node pool ${req.body.agentPoolName} to ${req.body.count}`
        };

    } catch (error) {
        context.res = {
            body: `Error: ${error}`
        };
    }

    context.done();

};`

The thing is if I use the same code to scale to more than 0 It returns a 200 but It does nothing...and I assume the same code It should work in both ways, am I right?

Thank you so much!

**Comment by btnguyen2k** [comment]
Anyone has a demo ARM template with user nodepool auto-scaled to 0?
I tried to deploy a nodepool using ARM template with `"minCount": 0` but got the error message:
```
{
    "code": "InvalidParameter",
    "message": "The value of parameter agentPoolProfile.minCount is invalid. Please see https://aka.ms/aks-naming-rules for more details.",
    "target": "agentPoolProfile.minCount"
}
```

**Comment by engineertree5** [comment]
> Correct, this should be available everywhere now for clusters with at least one system pool. Learn more at aka.ms/aks/nodepool/mode thanks for the feedback!
The Azure CLI will support this in a new version release in May. For now please use any clients such has ARM template until Azure CLI supports this capability.

I tried to scale my existing node pool 0 but that did not work. Az cli updated to the latest and greatest. I then created a system node pool to match what @jluk update said.. Again, tried scaling down that system node pool to zero and still was unable to do so. no 🎲 

having X number of nodepools scaled to zero, then running heavy computational batch jobs across said nodepools (model training) is the ideal use-case. Looking forward/waiting for this feature to hit GA so I can truly scale my node pool to 0! Once this happens i can like AzureDevops to deploy batch jobs to and scale up does to job requirements.

**Comment by jluk** [comment]
@engineertree5 only user node pools can be scaled to zero, the system node pool must have at least one node in it at all times, rely on this doc for details: aka.ms/aks/nodepool/mode

The CLI work is ongoing, so the earlier statement holds true for now that CLI isn't supported yet for scale to zero user pools.

@btnguyen2k support for the cluster autoscaler to allow a 0 node mincount is ongoing as well, we will clarify this on the doc mentioned above. We can use this ticket to track that. https://github.com/Azure/AKS/issues/1565

Thanks everyone for the feedback, we're on it.

**Comment by btnguyen2k** [comment]
@jluk thank you very much.
Just for clarification: scale node pool size to 0 is possible if autoscale is disabled. Because even we can manually change node pool size when autoscale=enabled but we can not set the size out of range `[minCount, maxCount]` and currently `minCount` must be a positive number. Am I correct?

**Comment by jluk** [comment]
That is correct

**Comment by msenmurugan** [comment]
@jluk There are requirements that we need to schedule an application specific jobs to be triggered in a particular node pool. And the jobs may not be triggered most of the time. So we need to have a node pool with minimum as 0 when autoscale is enabled. This feature have been available in GKE and EKS. Can you please let me know that why in AKS alone, only if autoscale is disable, you are allowing as to set the node pool size as 0?

**Comment by jluk** [comment]
The capability is being worked on as mentioned above for #1565 due to additional changes being required for the managed CA.  It is going to be supported, but requires more work.

In the interim there was no reason to block exposure of manual scale to 0 for a nodepool while we continued work on the CA 0 scale support, if for example someone wanted to cut cost on a nodepool setup on the weekends manually or utilize a custom autoscaling solution.

**Comment by msenmurugan** [comment]
Thanks @jluk for the update.

**Comment by 0GiS0** [comment]
> Hi all,
> 
> I wrote an Azure Function to try to scale to 0 using the following code and **It works perfectly**:
> 
> `var msRestNodeAuth = require('@azure/ms-rest-nodeauth');
> const msRestAzure = require('@azure/ms-rest-azure-js');
> 
> module.exports = async function (context, req) {
> 
> ```
> var creds = await msRestNodeAuth.loginWithServicePrincipalSecret(
>     process.env.CLIENT_ID,
>     process.env.SECRET,
>     process.env.TENANT_ID
> );
> 
> var client = new msRestAzure.AzureServiceClient(creds);
> 
> var options = {
>     method: 'PUT',
>     url: `https://management.azure.com/subscriptions/${req.body.subscriptionId}/resourceGroups/${req.body.resourceGroupName}/providers/Microsoft.ContainerService/managedClusters/${req.body.resourceName}/agentPools/${req.body.agentPoolName}?api-version=2020-03-01`,
>     body: {
>         "properties.count": req.body.count
>     }
> };
> 
> try {
>     var response = await client.sendRequest(options);
>     context.log(response);
>     context.log(`status code: ${response.status}`);
> 
>     context.res = {
>         body: `Node pool ${req.body.agentPoolName} to ${req.body.count}`
>     };
> 
> } catch (error) {
>     context.res = {
>         body: `Error: ${error}`
>     };
> }
> 
> context.done();
> ```
> 
> };`
> 
> The thing is if I use the same code to scale to more than 0 It returns a 200 but It does nothing...and I assume the same code It should work in both ways, am I right?
> 
> Thank you so much!

Hi again, 

any ideas about this? :(

Thanks again!

**Comment by akosbalazs** [comment]
> The Azure CLI will support this in a new version release in May. For now please use any clients such has ARM template until Azure CLI supports this capability.

@jluk Any update or ETA on the Azure CLI support? In the meantime could you post an ARM template examples for scaling to 0 (and back) ? Thanks.

**Comment by jluk** [comment]
CLI support has landed in the preview extension, please update and retry. I have confirmed it works.

Core CLI support has been merged and will release on the next release targeting 7/14.
https://github.com/Azure/azure-cli/pull/13996

**Comment by ohorvath** [comment]
> CLI support has landed in the preview extension, please update and retry. I have confirmed it works.
> 
> Core CLI support has been merged and will release on the next release targeting 7/14.
> [Azure/azure-cli#13996](https://github.com/Azure/azure-cli/pull/13996)

It seems scaling to zero node with autoscaler also works. :)

**Comment by akosbalazs** [comment]
> It seems my clusters can now scale down to zero with cluster autoscaler however the mincount is set to one.

Do you mean that a nodepool may autoscale down to 0 now? I'd be surprised because that feature (#1565) isn't ready yet and indeed when I try this command it doesn't let me set `--min-count` to 0 (even with the "preview extension" mentioned above):

```
az aks nodepool update --enable-cluster-autoscaler --min-count 0 --max-count 1 --resource-group foo --cluster-name bar  --name gpupool
The behavior of this command has been altered by the following extension: aks-preview
Operation failed with status: 'Bad Request'. Details: The value of parameter agentPoolProfile.minCount is invalid. Please see https://aka.ms/aks-naming-rules for more details.
```

If I set `--min-count` to 1 it will work but I don't think it will ever autoscale down to 0.

**Comment by ohorvath** [comment]
I'm not sure what's happening, but I had multiple clusters which scaled down to zero even with a min count set to 1. I didn't change my arm templates or anything. Of course I can't repro now, but I have screenshots. :)

**Comment by jluk** [comment]
@akosbalazs is correct, the support of CA scale to and from zero is still open and on track for delivery in the next 4-6 weeks as it requires adding support in the upstream CA project as well.

@ohorvath in the moment you should be able to manually scale to and from zero, but if CA is enabled when the node count is zero the scaling up functionality may not work. If you have a cluster entering a zero node count without you triggering that, it may be a bug and something we can look at via support ticket.

Hope this helps.

**Comment by btnguyen2k** [comment]
If you use spot nodepool, it can be scaled down to zero without you triggering it. I believe this case is valid (i.e. not bug).

**Comment by MuziSakura** [comment]
> @engineertree5 only user node pools can be scaled to zero, the system node pool must have at least one node in it at all times, rely on this doc for details: aka.ms/aks/nodepool/mode
> 
> The CLI work is ongoing, so the earlier statement holds true for now that CLI isn't supported yet for scale to zero user pools.
> 
> @btnguyen2k support for the cluster autoscaler to allow a 0 node mincount is ongoing as well, we will clarify this on the doc mentioned above. We can use this ticket to track that. #1565
> 
> Thanks everyone for the feedback, we're on it.

Hi, I try to add another system node pool and change previous system node pool to "User" mode successfully. But I still can't scale that previous node pool to zero. Does it seems right?

**Comment by jluk** [comment]
@btnguyen2k that is indeed a valid case, but once you hit zero nodes due to Spot eviction - you will need to manually request more nodes even if CA is enabled due to issue #1565. Once we close that CA should automatically scale up from zero as well if Spot nodes are available at your maxPrice.

@MuziSakura can you open a new issue with the error you are getting when trying to manually scale a user node pool? That is not expected, you should be able to manually scale to zero any user node pool.
